src/Microsoft.ML.TorchSharp/Roberta/QATrainer.cs

```diff
@@ -0,0 +1,938 @@
+                        targetTensors.TrimExcess();
+                        if (inputTensors.Count() == 0)
+                            return cursorValid;
+                    }
+                }
+
+                Updates++;
+                host.CheckAlive();
+
+                Optimizer.zero_grad();
+
+                srcTensor = PrepareBatchTensor(ref inputTensors, device: Device, Tokenizer.RobertaModel().PadIndex);
+                targetTensor = PrepareBatchTensor(ref targetTensors, device: Device, 0);
+
+                var logits = Model.forward(srcTensor);  //[batchsize, maxseqlen, 2]
+                var splitLogits = logits.split(1, dim: -1);
+                var startLogits = splitLogits[0].squeeze(-1).contiguous();  //[batchsize, maxseqlen]
+                var endLogits = splitLogits[1].squeeze(-1).contiguous();  //[batchsize, maxseqlen]
+
+                var targetsLong = targetTensor.@long();
+                var splitTargets = targetsLong.split(1, dim: -1);
+                var startTargets = splitTargets[0].squeeze(-1).contiguous();  //[batchsize]
+                var endTargets = splitTargets[1].squeeze(-1).contiguous();  //[batchsize]
+
+                torch.Tensor lossStart = torch.nn.CrossEntropyLoss(reduction: Parent.Option.Reduction).forward(startLogits, startTargets);
+                torch.Tensor lossEnd = torch.nn.CrossEntropyLoss(reduction: Parent.Option.Reduction).forward(endLogits, endTargets);
+
+                var loss = ((lossStart + lossEnd) / 2);
+
+                loss.backward();
+
+                Optimizer.Optimizer.step();
+                LearningRateScheduler.step();
+                host.CheckAlive();
+
+                if (Updates % Parent.Option.LogEveryNStep == 0)
+                {
+                    pch.Checkpoint(loss.ToDouble(), Updates);
+                    _channel.Info($"Row: {Updates}, Loss: {loss.ToDouble()}");
+                }
+
+                return cursorValid;
+            }
+
+            private torch.Tensor PrepareBatchTensor(ref List<Tensor> inputTensors, Device device, int padIndex)
+            {
+                return DataUtils.CollateTokens(inputTensors, padIndex, device: Device);
+            }
+
+            private (Tensor image, Tensor Label) PrepareData(ValueGetter<ReadOnlyMemory<char>> contextGetter, ValueGetter<ReadOnlyMemory<char>> questionGetter, ValueGetter<ReadOnlyMemory<char>> answerGetter, ValueGetter<int> answerIndexGetter)
+            {
+                using (var _ = torch.NewDisposeScope())
+                {
+                    ReadOnlyMemory<char> context = default;
+                    ReadOnlyMemory<char> question = default;
+                    ReadOnlyMemory<char> answer = default;
+                    int answerIndex = default;
+
+                    contextGetter(ref context);
+                    questionGetter(ref question);
+                    answerGetter(ref answer);
+                    answerIndexGetter(ref answerIndex);
```

Comment:
I'm not quite follow the `answerIndex` here,, especially why we need it as input.. Can you explain it a little bit

Reply:
You need it as an input for training, not for inferencing. I just also had it in the dataset and was pulling it out here.